src/Sulu/Component/CustomUrl/Manager/CustomUrlManager.php

```diff
@@ -0,0 +1,158 @@
+{
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $documentManager;
+
+    /**
+     * @var CustomUrlRepository
+     */
+    private $customUrlRepository;
+
+    /**
+     * @var PathBuilder
+     */
+    private $pathBuilder;
+
+    public function __construct(
+        DocumentManagerInterface $documentManager,
+        CustomUrlRepository $customUrlRepository,
+        PathBuilder $pathBuilder
+    ) {
+        $this->documentManager = $documentManager;
+        $this->customUrlRepository = $customUrlRepository;
+        $this->pathBuilder = $pathBuilder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create($webspaceKey, array $data, $locale = null)
+    {
+        $document = $this->documentManager->create('custom_urls');
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist(
+            $document,
+            $locale,
+            [
+                'parent_path' => $this->getItemsPath($webspaceKey),
+                'node_name' => Urlizer::urlize($document->getTitle()),
+                'load_ghost_content' => true,
+            ]
+        );
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function readList($webspaceKey)
+    {
+        // TODO pagination
+
+        return $this->customUrlRepository->findList($this->getItemsPath($webspaceKey));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read($uuid, $locale = null)
+    {
+        return $this->documentManager->find($uuid, $locale, ['load_ghost_content' => true]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update($uuid, array $data, $locale = null)
+    {
+        $document = $this->read($uuid, $locale);
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist($document, $locale);
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete($uuid)
+    {
+        $this->documentManager->remove($this->read($uuid));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFields()
+    {
+        return [
+            'title' => ['property' => 'title'],
+            'published' => ['property' => 'published'],
+            'baseDomain' => ['property' => 'baseDomain'],
+            'domainParts' => ['property' => 'domainParts', 'type' => 'json_array'],
+            'target' => ['property' => 'target', 'type' => 'reference'],
+            'multilingual' => ['property' => 'multilingual'],
+            'canonical' => ['property' => 'canonical'],
+            'redirect' => ['property' => 'redirect'],
+            'targetLocale' => ['property' => 'targetLocale'],
+        ];
+    }
+
+    /**
+     * Bind data array to given document.
+     *
+     * @param CustomUrlDocument $document
+     * @param array $data
+     * @param string $locale
+     */
+    private function bind(CustomUrlDocument $document, $data, $locale)
```

Comment:
However this is handled, This class is a repetition of the ContentMapper pattern.

The `read`, `update` and `remove` functions can be called from the Controller on the DM directly, and short of a dedicated layer, that is where I would put the request mapping also rather than providing a system-wide API which can be missused (we should not be encouraging this sort of thing), this would make refactoring easier in the future.

The `getFields` method should be handled by the DM Subscriber (as we have no way to load document metadata yet) and the `parent path` can be set I think via. the same mechanism we use for snippets etc.

In short, I don't think this class should exist.

This Manager class has multiple distinct responsibilities (binding, finding, writing, mapping) all of which should or are already handled by dedicated systems. At best this would be a pointless proxy class with an ambiguous name which is liable to missuse.